File: example2/zmq-word-count.cpp
#include <phish-bait.hpp>

int main(int argc, char* argv[])
{
  std::vector<std::string> files(argv + 1, argv + argc);
  std::vector<std::string> filegen_hosts(1, "localhost");
  std::vector<std::string> filegen_arguments(1, "zmq-filegen");
  filegen_arguments.insert(filegen_arguments.end(), files.begin(), files.end());
  std::vector<int> filegen_minnows = phish::bait::add_minnows("filegen", filegen_hosts, filegen_arguments);

  std::vector<std::string> file2words_hosts(1, "localhost");
  std::vector<std::string> file2words_arguments(1, "zmq-file2words");
  std::vector<int> file2words_minnows = phish::bait::add_minnows("file2word", file2words_hosts, file2words_arguments);

  std::vector<std::string> count_hosts(1, "localhost");
  std::vector<std::string> count_arguments(1, "zmq-count");
  std::vector<int> count_minnows = phish::bait::add_minnows("count", count_hosts, count_arguments);

  std::vector<std::string> sort_hosts(1, "localhost");
  std::vector<std::string> sort_arguments(1, "zmq-sort");
  sort_arguments.insert(sort_arguments.end(), "20"); // Return the top-twenty words
  std::vector<int> sort_minnows = phish::bait::add_minnows("sort", sort_hosts, sort_arguments);

  std::vector<std::string> print_hosts(1, "localhost");
  std::vector<std::string> print_arguments(1, "zmq-print");
  std::vector<int> print_minnows = phish::bait::add_minnows("print", print_hosts, print_arguments);

  phish::bait::all_to_all(filegen_minnows, 0, PHISH_BAIT_ROUND_ROBIN, 0, file2words_minnows);
  phish::bait::all_to_all(file2words_minnows, 0, PHISH_BAIT_HASHED, 0, count_minnows);
  phish::bait::all_to_all(count_minnows, 0, PHISH_BAIT_BROADCAST, 0, sort_minnows);
  phish::bait::one_to_one(sort_minnows, 0, 0, print_minnows);

  phish::bait::start();

  return 0;
}

File: example2/zmq-word-count.py
import optparse
importbait 
import sys

parser = optparse.OptionParser()
parser.add_option("--count", type="int", default=20, help="Number of items to return.  Default: %default.")
parser.add_option("--file2word-workers", type="int", default=1, help="Number of workers.  Default: %default.")
parser.add_option("--count-workers", type="int", default=1, help="Number of workers.  Default: %default.")
(options, arguments) = parser.parse_args()

filegen = bait.add_minnows("filegen", ["localhost"], ["zmq-filegen"] + arguments)
file2words = bait.add_minnows("file2words", ["localhost"] * options.file2word_workers, ["zmq-file2words"])
count = bait.add_minnows("count", ["localhost"] * options.count_workers, ["zmq-count"])
sort = bait.add_minnows("sort", ["localhost"], ["zmq-sort", str(options.count)])
print_messages = bait.add_minnows("print", ["localhost"], ["zmq-print"])

bait.all_to_all(filegen, 0, bait.ROUND_ROBIN, 0, file2words)
bait.all_to_all(file2words, 0, bait.HASHED, 0, count)
bait.all_to_all(count, 0, bait.BROADCAST, 0, sort)
bait.one_to_one(sort, 0, 0, print_messages)

bait.start()

